src/lava/magma/core/model/py/connection.py

```diff
@@ -0,0 +1,1702 @@
+                    trace_idx = 0
+                elif trace == str_symbols.X2:
+                    trace_idx = 1
+                elif trace == str_symbols.Y1:
+                    trace_idx = 2
+                elif trace == str_symbols.Y2:
+                    trace_idx = 3
+                elif trace == str_symbols.Y3:
+                    trace_idx = 4
+                else:
+                    raise ValueError("Unknown Trace in ProcessModel")
+
+                active_traces_per_dependency[dependency_idx, trace_idx] = True
+
+        # Shape : (3, 2)
+        self._active_x_traces_per_dependency = active_traces_per_dependency[
+                                               :, :2
+                                               ]
+
+        # Shape : (3, 3)
+        self._active_y_traces_per_dependency = active_traces_per_dependency[
+                                               :, 2:
+                                               ]
+
+    def _build_active_traces(self) -> None:
+        """Build and store boolean numpy arrays specifying which x and y
+        traces are active."""
+        # Shape : (2, )
+        self._active_x_traces = np.logical_or(
+            self._active_x_traces_per_dependency[0],
+            self._active_x_traces_per_dependency[1],
+            self._active_x_traces_per_dependency[2],
+        )
+
+        # Shape : (3, )
+        self._active_y_traces = np.logical_or(
+            self._active_y_traces_per_dependency[0],
+            self._active_y_traces_per_dependency[1],
+            self._active_y_traces_per_dependency[2],
+        )
+
+    def _build_learning_rule_appliers(self) -> None:
+        """Build and store LearningRuleApplierBitApprox for each active learning
+        rule in a dict mapped by the learning rule's target."""
+        self._learning_rule_appliers = {
+            str_symbols.SYNAPTIC_VARIABLE_VAR_MAPPING[
+                target[1:]
+            ]: LearningRuleApplierBitApprox(ps)
+            for target, ps in self._learning_rule.active_product_series.items()
+        }
+
+    def _init_randoms(self) -> None:
+        """Initialize trace and synaptic variable random generators."""
+        self._x_random = \
+            TraceRandom(seed_trace_decay=self._learning_rule._rng_seed,
+                        seed_impulse_addition=self._learning_rule._rng_seed + 1)
+
+        self._y_random = \
+            TraceRandom(seed_trace_decay=self._learning_rule._rng_seed + 2,
+                        seed_impulse_addition=self._learning_rule._rng_seed + 3)
+
+        self._conn_var_random = ConnVarRandom()
+
+    def _update_trace_randoms(self) -> None:
+        """Update trace random generators."""
+        self._x_random.advance()
+        self._y_random.advance()
+
+    def _update_synaptic_variable_random(self) -> None:
+        """Update synaptic variable random generators."""
+        self._conn_var_random.advance()
+
+    # Shape: (2, num_pre_neurons)
```

Comment:
Move this into the docstring under `Returns`.

Reply:
Same here, we restructured the code for more readability.